pkg/event/event.go

```diff
@@ -133,13 +133,24 @@ func WithAck(handledFn func(Event), errFn func(Event, error)) EventOption {
 	}
 }
 
-func WithMetadata(metadata interface{}) EventOption {
+func WithMetadata(metadata map[string]interface{}) EventOption {
 	return func(e *event) error {
 		e.SetMetadata(metadata)
 		return nil
 	}
 }
 
+func WithMetadataKeyValue(key string, val interface{}) EventOption {
+	return func(e *event) error {
+		path := key
+		if !strings.HasPrefix(path, METADATA+".") {
+			path = METADATA + "." + path
+		}
+		e.SetPathValue(path, val, true)
```

Comment:
I found this code confusing. Path is a superset of metadata, payload, tenant, etc. SetPathValue function should be calling SetMedatadataKeyValue instead of the other way around. Also, SetPathValue is not balanced between Set and Get (ie, it seems that you can get tenant from path but you cannot set it?). I guess I will be happier if Set/GetPathValue is not part of event functions. Instead, they are helper functions that use the Set/Get Metadata, and Set/Get Payload to manipulate the internals of an event.

Reply:
understood - privatizing Set/Get path value would be a bigger change though - I think they are used in several places for different reasons. Also the whole business of path value syntax is part of our user guide etc.

Reply:
I am not asking that we privatize get/set path value. I am pointing out that path is not really a *thing* in the event. Instead, it is a derived value from event payload/metadata/tenant/etc. Path is higher level than payload/metadata, so that the get/set path functions should be calling get/set metadata & get/set payload to manipulate these values. Functionally, there is nothing wrong with what you are doing here, but I am saying for code organization, it is weird that when we are setting metadata (which is more basic than path), it needs to call setpath function instead of just manipulating the metadata fields directly.

Reply:
ok, i changed the implementation per your comment